File: library/ilib/ml/ffi/ml_ffi.js
const ml = (() => {
  let nextId = 1;
  const nodes = new Map();
  const put = (value, op = '', a = 0, b = 0) => {
    const id = nextId++;
    nodes.set(id, { value: Number(value), grad: 0, track: false, op, a, b });
    return id;
  };
  const node = (id) => nodes.get(Number(id));
  const take = (id) => node(id)?.value ?? 0;
  const addGrad = (id, g) => { const n = node(id); if (n) n.grad += g; };
  const back = (id, upstream) => {
    const n = node(id);
    if (!n) return;
    n.grad += upstream;
    if (n.op === 'add') {
      back(n.a, upstream);
      back(n.b, upstream);
    } else if (n.op === 'mul') {
      back(n.a, upstream * take(n.b));
      back(n.b, upstream * take(n.a));
    } else if (n.op === 'relu' && take(n.a) > 0) {
      back(n.a, upstream);
    }
  };
  return {
    tensor(value) { return put(value); },
    grid(rows, cols, fill) { return put(fill); },
    gradient_track(tensorId, enabled) { const n = node(tensorId); if (!n) return 0; n.track = Number(enabled) !== 0; return 1; },
    grad_wipe(tensorId) { const n = node(tensorId); if (!n) return 0; n.grad = 0; return 1; },
    weights(learningRate, tensorId) { const n = node(tensorId); if (!n) return 0; n.value -= Number(learningRate) * n.grad; return 1; },
    take,
    add(a, b) { return put(take(a) + take(b), 'add', Number(a), Number(b)); },
    multiply(a, b) { return put(take(a) * take(b), 'mul', Number(a), Number(b)); },
    relu(x) { return put(Math.max(0, take(x)), 'relu', Number(x), 0); },
    backward(loss) { if (!node(loss)) return 0; back(Number(loss), 1); return 1; },
    // ml.acl maps ml:grad -> ml.grad / ml:optimize -> ml.optimize — both were entirely
    // missing here (matches ml_get_grad/ml_sgd_step's own naming gap fixed in ml.cpp/every
    // other binding). grad() mirrors native's ml_get_grad: a NEW tensor node holding a copy
    // of the accumulated gradient, not a live view. optimize() is a plain alias of weights().
    grad(tensorId) { const n = node(tensorId); return put(n ? n.grad : 0); },
    optimize(learningRate, tensorId) { const n = node(tensorId); if (!n) return 0; n.value -= Number(learningRate) * n.grad; return 1; },
  };
})();

File: library/ilib/ml/ffi/ml_ffi.java
class ml {
    private static class Node {
        double value;
        double grad;
        boolean track;
        String op;
        long a;
        long b;

        Node(double value, String op, long a, long b) {
            this.value = value;
            this.op = op;
            this.a = a;
            this.b = b;
        }
    }

    private static long nextId = 1;
    private static final java.util.Map<Long, Node> nodes = new java.util.HashMap<>();

    private static long put(double value, String op, long a, long b) {
        long id = nextId++;
        nodes.put(id, new Node(value, op, a, b));
        return id;
    }

    private static Node node(long id) { return nodes.get(id); }

    private static void back(long id, double upstream) {
        Node n = node(id);
        if (n == null) return;
        n.grad += upstream;
        if ("add".equals(n.op)) {
            back(n.a, upstream);
            back(n.b, upstream);
        } else if ("mul".equals(n.op)) {
            back(n.a, upstream * take(n.b));
            back(n.b, upstream * take(n.a));
        } else if ("relu".equals(n.op) && take(n.a) > 0.0) {
            back(n.a, upstream);
        }
    }

    static long tensor(double value) { return put(value, "", 0, 0); }
    static long grid(long rows, long cols, double fill) { return put(fill, "", 0, 0); }
    static long gradient_track(long tensorId, long enabled) {
        Node n = node(tensorId);
        if (n == null) return 0;
        n.track = enabled != 0;
        return 1;
    }
    static long grad_wipe(long tensorId) {
        Node n = node(tensorId);
        if (n == null) return 0;
        n.grad = 0.0;
        return 1;
    }
    static long weights(double learningRate, long tensorId) {
        Node n = node(tensorId);
        if (n == null) return 0;
        n.value -= learningRate * n.grad;
        return 1;
    }
    static double take(long tensorId) {
        Node n = node(tensorId);
        return n == null ? 0.0 : n.value;
    }
    static long add(long a, long b) { return put(take(a) + take(b), "add", a, b); }
    static long multiply(long a, long b) { return put(take(a) * take(b), "mul", a, b); }
    static long relu(long x) { return put(Math.max(0.0, take(x)), "relu", x, 0); }
    static long backward(long loss) {
        if (!nodes.containsKey(loss)) return 0;
        back(loss, 1.0);
        return 1;
    }
}
